### PR TITLE
[FLINK-31597][table] Cleanup usage of deprecated TableEnvironment#registerFunction

### PR DESCRIPTION
**Pull request by snuyanzin**
## What is the purpose of the change

Removes the usage of the deprecated method `TableEnvironment#registerFunction`. 

The PR is based on https://github.com/apache/flink/pull/23748

## Brief change log

- Replace the usage of `TableEnvironment#registerFunction` with `TableEnvironment#createTemporarySystemFunction`

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by snuyanzin** [comment]
@wuchong since you were involved in this, could you please have a look?

It seems I found a way to clean this deprecated things

**Comment by wuchong** [comment]
cc @lincoln-lil , @LadyForest , could you help to review this?

**Comment by LadyForest** [comment]
> cc @lincoln-lil , @LadyForest , could you help to review this?

Sure. I'll take a look ASAP.

**Comment by snuyanzin** [comment]
the failure is unrelated and the reason is FLINK-33568

**Comment by snuyanzin** [comment]
@flinkbot run azure

**Review by LadyForest — COMMENTED** [review]
Hi @snuyanzin, thanks for the contribution. The PR looks good to me, and I just left some minor comments.

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/DagOptimizationTest.scala [hunk 1]
```diff
@@ -235,7 +235,7 @@ class DagOptimizationTest extends TableTestBase {
       RelNodeBlockPlanBuilder.TABLE_OPTIMIZER_REUSE_OPTIMIZE_BLOCK_WITH_DIGEST_ENABLED,
       Boolean.box(true))
     // test with non-deterministic udf
-    util.tableEnv.registerFunction("random_udf", new NonDeterministicUdf())
+    util.tableEnv.createTemporarySystemFunction("random_udf", new NonDeterministicUdf())
```
[the remark this answers is not in the document]
Comment (snuyanzin) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/DagOptimizationTest.scala]: initially I just followed javadoc saying that `registerFunction` should be replaced by `createTemporarySystemFunction` however agree in this case we could use `addTemporarySystemFunction`

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/OverAggregateITCase.scala [hunk 2]
```diff
@@ -22,14 +22,16 @@ import org.apache.flink.api.common.typeinfo.TypeInformation
 import org.apache.flink.api.java.tuple.{Tuple1 => JTuple1}
 import org.apache.flink.api.java.typeutils.{RowTypeInfo, TupleTypeInfo}
 import org.apache.flink.api.scala._
-import org.apache.flink.table.api.Types
+import org.apache.flink.table.api.{DataTypes, Types}
```
[the remark this answers is not in the document]
Comment (snuyanzin) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/OverAggregateITCase.scala]: yeah.... unfortunately `spotless:apply` doesn't fix scala imports... 
fix it with IDEA's functionality

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/utils/UserDefinedFunctionTestUtils.scala [hunk 3]
```diff
@@ -275,21 +296,44 @@ object UserDefinedFunctionTestUtils {
...
   }
 
   @SerialVersionUID(1L)
   object MyToPojoFunc extends ScalarFunction {
-    def eval(s: Int): MyPojo = new MyPojo(s, s)
 
-    override def getResultType(signature: Array[Class[_]]): PojoTypeInfo[MyPojo] = {
+    def eval(s: Int) = new MyPojo(s, s)
+
+    /*override def getResultType(signature: Array[Class[_]]): PojoTypeInfo[MyPojo] = {
```
[the remark this answers is not in the document]
Comment (snuyanzin) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/utils/UserDefinedFunctionTestUtils.scala]: done

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/utils/CountAggFunction.scala [hunk 4]
```diff
@@ -20,8 +20,11 @@ package org.apache.flink.table.planner.utils
 import org.apache.flink.api.common.typeinfo.{BasicTypeInfo, TypeInformation}
 import org.apache.flink.api.java.tuple.{Tuple1 => JTuple1}
 import org.apache.flink.api.java.typeutils.TupleTypeInfo
-import org.apache.flink.table.api.Types
+import org.apache.flink.table.annotation.DataTypeHint
+import org.apache.flink.table.api.{DataTypes, Types}
```
[the remark this answers is not in the document]
Comment (snuyanzin) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/utils/CountAggFunction.scala]: done

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/utils/SumAggFunction.scala [hunk 5]
```diff
@@ -20,7 +20,13 @@ package org.apache.flink.table.planner.utils
 import org.apache.flink.api.common.typeinfo.{BasicTypeInfo, TypeInformation}
 import org.apache.flink.api.java.tuple.{Tuple2 => JTuple2}
 import org.apache.flink.api.java.typeutils.TupleTypeInfo
+import org.apache.flink.table.annotation.{DataTypeHint, FunctionHint}
```
[the remark this answers is not in the document]
Comment (snuyanzin) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/utils/SumAggFunction.scala]: done

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/CorrelateITCase.scala [hunk 6]
```diff
@@ -23,8 +23,12 @@ import org.apache.flink.api.java.typeutils.{PojoField, PojoTypeInfo, RowTypeInfo
 import org.apache.flink.api.scala._
 import org.apache.flink.api.scala.typeutils.Types
 import org.apache.flink.configuration.Configuration
+import org.apache.flink.table.annotation.{DataTypeHint, FunctionHint, HintFlag}
```
[the remark this answers is not in the document]
Comment (snuyanzin) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/batch/sql/CorrelateITCase.scala]: done

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/utils/UserDefinedTableFunctions.scala [hunk 7]
```diff
@@ -429,12 +431,24 @@ abstract class SplittableTableFunction[A, B] extends TableFunction[Tuple3[String
...
   }
+
+  override def getTypeInference(typeFactory: DataTypeFactory): TypeInference = {
+    TypeInference.newBuilder
+      .typedArguments(DataTypes.STRING())
+      .outputTypeStrategy(
+        TypeStrategies.explicit(
+          DataTypes.STRUCTURED(
+            classOf[PojoUser],
+            DataTypes.FIELD("age", DataTypes.INT()),
```
[the remark this answers is not in the document]
Comment (snuyanzin) [reply flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/utils/UserDefinedTableFunctions.scala]: done

File: flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/stream/table/CorrelateTest.xml [hunk 8]
```diff
@@ -321,12 +321,12 @@ Correlate(invocation=[*org.apache.flink.table.planner.utils.TableFunc1*(SUBSTRIN
       <![CDATA[
 LogicalCorrelate(correlation=[$cor0], joinType=[inner], requiredColumns=[{}])
 :- LogicalTableScan(table=[[default_catalog, default_database, MyTable, source: [TestTableSource(a, b, c)]]])
-+- LogicalTableFunctionScan(invocation=[*org.apache.flink.table.planner.utils.PojoTableFunc*($2)], rowType=[RecordType(VARCHAR(2147483647) name, INTEGER age)])
++- LogicalTableFunctionScan(invocation=[*org.apache.flink.table.planner.utils.PojoTableFunc*($2)], rowType=[RecordType(INTEGER age, VARCHAR(2147483647) name)])
```
[the remark this answers is not in the document]
Comment (snuyanzin) [reply flink-table/flink-table-planner/src/test/resources/org/apache/flink/table/planner/plan/stream/table/CorrelateTest.xml]: done

**Review by LadyForest — APPROVED** [review]
Thanks for the update, looks good to me :)

**Comment by LadyForest** [comment]
I just noticed that the master branch has been updated, would you mind rebasing and resolving conflicts? @snuyanzin

**Comment by snuyanzin** [comment]
Thanks for notifying about that, resolved the conflicts

I'm going to merge it once ci is passed

i'm going to merge as 2 commits as each commit is related to a different jira issue

**Review by JingGe — APPROVED** [review]
LGTM

**Comment by snuyanzin** [comment]
thanks for taking a look

**Pull request merged**